ext/socket/lib/socket.rb

```diff
@@ -593,6 +594,93 @@ def accept_nonblock(exception: true)
     __accept_nonblock(exception)
   end
 
+  def self.start_getaddrinfo_v6(host, port, queue, resolv_timeout)
+    return Thread.new do
+      ai_list = Addrinfo.getaddrinfo(host, port, :PF_INET6, :STREAM, timeout: resolv_timeout)
+      ai_list.each {|ai| queue.push(ai) }
+    rescue SocketError => ex
+      case ex.message
+      when "getaddrinfo: Address family for hostname not supported" # when IPv6 is not supported
+        # ignore
+      when "getaddrinfo: Temporary failure in name resolution" # when timed out (EAI_AGAIN)
+        # ignore
+      else
+        raise ex
+      end
+    rescue ClosedQueueError
+      # timed out. ignore exception
+    ensure
+      begin
+        queue.push(:getaddrinfo_v6_done)
+      rescue ClosedQueueError
+        # timed out. ignore exception
+      end
+    end
+  end
+  private_class_method :start_getaddrinfo_v6
+
+  # 50ms is the recommended value for the resolution delay for IPv4 in RFC8305
+  RESOLUTION_DELAY = 0.05
+  private_constant :RESOLUTION_DELAY
+
+  def self.start_getaddrinfo_v4(host, port, queue, resolv_timeout)
+    return Thread.new do
+      ai_list = Addrinfo.getaddrinfo(host, port, :PF_INET, :STREAM, timeout: resolv_timeout)
+      sleep(RESOLUTION_DELAY) # 50ms is the recommended value for the resolution delay for IPv4 in RFC8305
```

Comment:
The resolution delay is applied here unconditionally.
I think it is not proper implementation of RFC 8305.

When IPv6 getaddrinfo returns a negative response immediately and
IPv4 getaddrinfo returns a positive response immediately,
this method should start to connect the IPv4 address without waiting the resolution delay.

Reply:
I fixed it to start to connect via IPv4 immediately if IPv6 getaddrinfo returns a negative response .